prusti-tests/tests/cargotest.rs

```diff
@@ -0,0 +1,170 @@
+
+To learn more, run the command again with --verbose.
+",
+        )
+        .run();
+}
+
+/// Test `cargo-prusti` on one of the crates in `test/cargo_verify`.
+///
+/// Special files and folders in the root of the test crate:
+/// * `output.stdout` and `output.stderr`: if present, they are used to check the output of
+///   `cargo-prusti`.
+/// * `prusti-contracts` and related Prusti crates: during the test they will link to the
+///   corresponding Prusti crate.
+///
+/// This function requires symlinks to be supported.
+///
+/// For more details on the special syntax allowed in the `output.*` files, check the documentation
+/// of `cargo_test_support`: <https://doc.crates.io/contrib/tests/writing.html>.
+fn test_local_project<T: Into<PathBuf>>(project_name: T) {
+    let mut project_builder = project().no_manifest();
+    let relative_project_path = Path::new("tests/cargo_verify").join(project_name.into());
+    let project_path = fs::canonicalize(&relative_project_path).expect(
+        &format!("Failed to canonicalize the path {}", relative_project_path.display())
+    );
+
+    // Populate the test project with symlinks to the local project
+    let project_path_content = fs::read_dir(&project_path)
+        .expect(&format!("Failed to read directory {}", project_path.display()));
+    for entry in project_path_content {
+        let entry = entry.expect(&format!("Failed to read content of {}", project_path.display()));
+        let path = entry.path();
+        let file_name = path.as_path().file_name()
+            .expect(&format!("Failed to obtain the name of {}", path.display()));
+        if path.is_dir() {
+            eprintln!("Symlink dir {:?} -> {}", file_name, path.display());
+            project_builder = project_builder.symlink_dir(path.as_path(), &Path::new(file_name));
+        } else {
+            eprintln!("Symlink file {:?} -> {}", file_name, path.display());
+            project_builder = project_builder.symlink(path.as_path(), &Path::new(file_name));
+        }
+    }
+
+    // Create a special symlink for prusti_contract and related Prusti crates
+    let prusti_dev_path = project_path
+        .parent()
+        .and_then(|p| p.parent())
+        .and_then(|p| p.parent())
+        .and_then(|p| p.parent())
+        .expect(&format!("Failed to obtain parent folders of {}", project_path.display()));
+    let prusti_contract_deps = [
+        "prusti-specs",
+        "prusti-contracts",
+        "prusti-contracts-impl",
+        "prusti-contracts-internal",
+    ];
+    for crate_name in &prusti_contract_deps {
+        eprintln!("Symlink dir {:?} -> {}", crate_name, prusti_dev_path.join(crate_name).display());
+        project_builder = project_builder.symlink_dir(
+            prusti_dev_path.join(crate_name).as_path(),
+            &Path::new(crate_name)
+        );
+    }
+
+    // Fetch dependencies
+    let project = project_builder.build();
+    project.process("cargo").arg("build").run();
```

Comment:
Hi @fpoli , I am trying to add a test case for `#[extern_spec]` with cargo-prusti, the attribute `#[extern_spec]` requires a rewrite for all the functions in the declared module but it will error out if we build the project with `cargo` first:
```rust
// Report an error because function is without body
#[extern_spec]
mod std {
    mod mem {
        use prusti_contracts::*;

        #[ensures(*a == old(*b) && *b == old(*a))]
        pub fn swap(a: &mut i32, b: &mut i32);
        // pub fn swap<T: std::cmp::PartialEq + Copy>(a: &mut T, b: &mut T);
    }
}
```

Wonder if we actually need this line or is it for compiling the dependencies only? Also, do you run cargo-prusti on the test case to see what the actual output is? I am having problem matching my expect output with the actual one

Reply:
@LDY1998 As we discussed in the meetings, this is something that you need to solve. The reason `cargo build` is executed first is to compile all dependencies and to make sure that both the dependencies and the main crate itself can be built with the regular, non-Prusti compiler. If a regular `cargo build` cannot build a crate that uses Prusti, it is a problem.

`cargo prusti` [is executed](https://github.com/viperproject/prusti-dev/blob/b58ced8dfd14ef30582b503d517167ccd771eaff/prusti-tests/tests/cargotest.rs#L137) too, and the stderr and stdout are compared.

Reply:
Exactly as Aurel wrote: the `cargo build` is needed to test that the crate compiles with the regular compiler, then `cargo-prusti` runs too. To reproduce the test it should be enough to run the cargo commands in the crate of the test, creating the symlinks for `prusti-contracts` and similar creates as the testsuite does.

Reply:
Sorry if I misunderstand it, the syntax for `#[extern_spec]` cannot be compiled with cargo since all the fns we specify there don't have body, it will produce an error and fail when we run cargo on the crate. So do we want to make this syntax work? Or is it really necessary to run `cargo build` since `cargo-prusti` will check the crate to see if it can be compiled?

Reply:
Yes, it is necessary that any crates using Prusti can also compile with the regular Rust compiler.

Please double check what code you are giving the compiler (e.g. with [`print_desugared_specs`](https://viperproject.github.io/prusti-dev/dev-guide/config/arguments.html)). `#[extern_spec]` on its own compiles to valid Rust code. You need to make sure that this processing still happens.

Reply:
I think I got it now:

`cargo build` will do the path resolution of the attribute, but the attribute is not called to actually make the rewrite happens, it just makes sure that the attribute is somewhere defined in crate/dependencies.

As the rewrite is not triggered, the syntax for `#[extern_spec]` will be a function without body, hence it will error out so that `cargo` cannot actually build it. 

The `print_desugared_specs` won't be able to trigger here since we don't even reach prusti, the process is stopped when we are building with `cargo`. 

The only way to get current syntax pass `cargo` is to have the user write some stub body for the function (which is not what we want), or I am wrong and we can actually get the attribute rewrite happen during `cargo build` without prusti